fix: Only store owned safes for their given chainId

usame-algan opened the pull request:
## What it solves

Resolves #1759 

## How this PR fixes it

- Awaits the fetch of owned safes before setting the cache to avoid setting it for a stale state

## How to test it

1. Go to Safe
2. Clear the ownedSafes local storage
3. Connect a wallet that owns safes on different networks
4. Open the safe list sidebar
5. Observe owned safes are fetched and saved in the local storage
6. Switch to a safe on a different network
7. Observe the safe list sidebar closes
8. Observe the `ownedSafes` local storage doesn't update
9. Open the safe list sidebar again
10. Observe that the `ownedSafes` local storage gets overriden with the correct entries

## Checklist
* [x] I've tested the branch on mobile 📱
* [x] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

francovenica commented:
The "Owned safes" are not being fetched on the sidebar after connecting with an owner. This also happens after switching networks, where the sidebar should populate with the safes owned on that network

Try to connect with a safe that owns safes. See the sidebar remains empty:
![image](https://user-images.githubusercontent.com/4503659/224594948-f554fc7a-397a-42df-b44c-95cecbc9d749.png)

usame-algan commented:
> Try to connect with a safe that owns safes. See the sidebar remains empty:

Nice catch, thanks for testing! It was because I was relying on the `SafeInfo` which doesn't exist on the welcome page. I reverted that change and fixed it another way.

katspaugh commented on src/hooks/useOwnedSafes.ts:
```diff
@@ -16,26 +16,23 @@ type OwnedSafesCache = {
-    if (!ownedSafes || !walletAddress || !safe.chainId) return
+    const ownedSafes = await getOwnedSafes(chainId, walletAddress)
 
     setOwnedSafesCache((prev) => ({
```
Setting the state inside useAsync is IMHO an anti-pattern.
Is there another way to fix it?

katspaugh replied:
What's the root cause of the bug? The chainId changes over time?

katspaugh replied:
If you want to keep it in this imperative style, I would suggest just using an effect like so:
```
useEffect() => {
  getOwnedSafes(chainId).then(data => setOwnedSafesCache(prev => {
    ...
  }))
}, [chainId, etc])
```

usame-algan replied:
The root cause is that the `Drawer` unmounts when switching to a safe on a different network but there is a brief period before it unmounts where the `chainId` already changed but the `ownedSafes` data is stale. It runs the `useEffect` hook which updates the local storage for the new `chainId` with the data from the previous network.

Alternatively, we could keep the `Drawer` mounted through the `keepMounted` prop that MUI offers which solves the issue but @iamacook raised a valid point that this would not update a newly created safe unless the user either reconnects their wallet or switches to a different chain.

Another thing we could do is to not use `useAsync` for this case to avoid this anti-pattern.

katspaugh commented on src/hooks/useOwnedSafes.ts:
```diff
@@ -16,26 +16,23 @@ type OwnedSafesCache = {
-    if (!safe.chainId || !walletAddress) return
-    return getOwnedSafes(safe.chainId, walletAddress)
-  }, [safe.chainId, walletAddress])
+  const [_, error] = useAsync<void>(async () => {
```
Syntax tip:
```suggestion
  const [, error] = useAsync<void>(async () => {
```
(Irrelevant tho, see my other comment).
